llama_index/postprocessor/rankGPT_rerank.py

```diff
@@ -0,0 +1,223 @@
+        self, messages: Sequence[ChatMessage], model: str = "gpt-3.5-turbo-0301"
+    ) -> int:
+        """Returns the number of tokens used by a list of messages."""
+        if model == "gpt-3.5-turbo":
+            return self._num_tokens_from_messages(messages, model="gpt-3.5-turbo-0301")
+        elif model == "gpt-4":
+            return self._num_tokens_from_messages(messages, model="gpt-4-0314")
+        elif model == "gpt-3.5-turbo-0301":
+            tokens_per_message = (
+                4  # every message follows <|start|>{role/name}\n{content}<|end|>\n
+            )
+            # tokens_per_name = -1  # if there's a name, the role is omitted
+        elif model == "gpt-4-0314":
+            tokens_per_message = 3
+            # tokens_per_name = 1
+        else:
+            tokens_per_message, tokens_per_name = 0, 0
+
+        try:
+            encoding = tiktoken.get_encoding(model)
+        except Exception as e:
+            logger.warning(
+                f"Encountered exception when loading encoding for given model: {e}"
+            )
+            encoding = tiktoken.get_encoding("cl100k_base")
+
+        num_tokens = 0
+        for message in messages:
+            num_tokens += tokens_per_message
+            if message.content is not None:
+                num_tokens += len(encoding.encode(str(message.content)))
+        num_tokens += 3  # every reply is primed with <|start|>assistant<|message|>
+        return num_tokens
+
+    def create_permutation_instruction(self, item: Dict[str, Any]) -> List[ChatMessage]:
+        query = item["query"]
+        num = len(item["hits"])
+
+        max_length = 300
+        while True:
+            messages = self._get_prefix_prompt(query, num)
+            rank = 0
+            for hit in item["hits"]:
+                rank += 1
+                content = hit["content"]
+                content = content.replace("Title: Content: ", "")
+                content = content.strip()
+                # For Japanese should cut by character: content = content[:int(max_length)]
+                content = " ".join(content.split()[: int(max_length)])
+                messages.append(ChatMessage(role="user", content=f"[{rank}] {content}"))
+                messages.append(
+                    ChatMessage(role="assistant", content=f"Received passage [{rank}].")
+                )
+            messages.append(
+                ChatMessage(role="user", content=self._get_post_prompt(query, num))
+            )
+            for m in messages:
+                print(m)
+            if (
+                self._num_tokens_from_messages(messages, self.model)
+                <= openai_modelname_to_contextsize(self.model) - 200
+            ):
+                break
+            else:
+                max_length -= 1
+        return messages
+
+    def run_llm(self, messages: Sequence[ChatMessage]) -> ChatResponse:
+        return OpenAI(temperature=0, model=self.model, api_key=self.api_key).chat(
```

Comment:
hardcoded to openai?

Reply:
yeah right now it mostly supports openai models. the name already refer it is gpt related

Reply:
but since you're implementing it, why can't we support other models?

Reply:
seems more intuitive for me to let the user pass in an optional llm param and have that default to openai

Reply:
@jerryjliu done the change. pls stamp